Callout enhancements [test ui]

cpirich opened the pull request:
- Added a new `codeStudio` object under the `qtip_config` object for callouts, where we can store custom state for each callout.
- Added new properties `canReappear` and `dropletPaletteCategory` under that `codeStudio` object and updated the level builder editor to make it easy to edit these. Extended the `jsonEditor` to handle boolean properties as checkboxes.
- App code can now detect that a callout is about to appear and make changes (e.g. toggling UI state or scrolling an item into view) prior to the callout appearing. A `prepareforcallout` event is fired with the aforementioned `codeStudio` object as a parameter after the event object.
- Handled the `prepareforcallout` event for callouts that specify the `dropletPaletteCategory`. The droplet palette will change to the proper group and the specific block will scroll into view.
- Fire new `droplet_change` events from the apps world so that the callout code in code-studio can make appropriate changes related to callouts. The additional parameter on this type of event can be `togglepalette`, `scrollpalette`, or `selectpalette`. This is analogous to how `block_space_metrics_set` events are handled that are fired by blockly core.
- Changed how "seen" callouts are tracked. The dashboard will send "already seen" callouts if the `canReappear` property is set, but they will contain a `seen` property that is set to `true`
  - Modified `showOrHideCalloutsByTargetVisibility` so it can be used for callouts attached to the droplet palette in addition to tracking blocks within the blockly workspace. This is now a function that generates two different functions: `showHideWorkspaceCallouts()` and `showHidePaletteCallouts()`. That function has been extended to notice when the callout selector is no longer part of the document, in which case it destroys the callout.
- Added code to handle `#triggercallout=<event name>` in the page URL. If it is there on page load, the event name is fired with `hashinit` as the parameter, which causes the callout with the designated event name to appear immediately (using the existing callout `on` property). If a navigation to this relative URL occurs on the same page (e.g. a link from within instructions markdown), the event is fired with `hashchange` as the parameter, causing the callout to appear and the browser history is altered so this doesn't count as a navigation.

cpirich commented:
@breville

breville commented:
fyi @bcjordan and @Bjvanminnen who probably know callouts better than moi.

cpirich commented:
FYI: the only test failure at this point is the UI test for `Feature: After completing the Hour of Code, the player is directed to a congratulations page` - which I believe to be a test issue unrelated to these changes.

Bjvanminnen commented on apps/src/StudioApp.js:
```diff
@@ -2116,10 +2116,31 @@ StudioApp.prototype.handleEditCode_ = function (config) {
+      }
+    }
+
+  }.bind(this));
```
note: could use ES6 arrow funcs and avoid the need to bind this, i.e.

```
$(window).on('prepareforcallout', (e, options) => {
   ...
  this.foo();
});
```

could also replace `var`'s with `const`'s

Bjvanminnen commented on code-studio/src/js/callouts.js:
```diff
@@ -112,16 +124,49 @@ module.exports = function createCallouts(callouts) {
+
+  // Insert a hashchange handler to detect triggercallout= hashes and fire
+  // appropriate events to open the callout
+  function detectTriggerCalloutOnHash(event) {
```
Would it make more sense to use query params (?triggercallout=foo) instead of hashes, similar to how we do for show_callouts?

Bjvanminnen replied:
What is triggercallout used for? I'm not sure I entirely understand.

cpirich replied:
The `#triggercallout=` is intended to be used by markdown in the instructions to force callouts to appear. I've attached an example video. Note how the droplet palette category changes automatically, the block scrolls into view, and the callouts now track the block as the palette scrolls.

The reason it doesn't use query parameters is so that it doesn't result in a trip back to the server - this is intended to be a client-side only URL change. But by using a standard URL format, it makes it easy to create links inside markdown.

Bjvanminnen replied:
Ok, I think I understand what you're doing now. I don't see a video anywhere though. If you do have one I wouldn't mind seeing it to confirm my understanding :)

Bjvanminnen replied:
Another way we could accomplish this (I think) would be to have levelbuilders add a data-attribute to the markup, so you'd have something like `<a href="#" data-triggercallout="something"/>See something callout</a>`. Then, in our Instructions component, after mounting we can look for any descendant links with this attribute and add onclick handlers.

Right now it seems like you're essentially using the URL hash as global mechanism for specifying event parameters, because that makes it easier to pass those parameters from markup. This feels hackier than I'd like.

cpirich replied:
So authoring in my current proposal looks like this:

```
You need to [draw sprites](#triggercallout=drawSpritesCallout) and
```

How does one author that URL above in markdown? I assume you'd have to use the more traditional HTML syntax and not markdown, correct? I'm not opposed to your change, but it may make the authoring harder in order to clean up the code on a little. But since we're not using hash tags for anything yet (many "apps" on the web use it extensively with full paths after the hash), I'm not sure it creates much of a problem.
Also, how does one attach to the link event in this case?

Bjvanminnen replied:
Ahh, writing that markup is even easier than I realized. I didn't realize you didnt need to explicitly define an anchor tag as is. Might be the tradeoff in ease is worth it. Let me spend a little bit more time thinking about this, and I'll respond to this PR again in a bit.

Bjvanminnen replied:
What is the scenario where the hash will be there on page load?

cpirich replied:
There isn't one now, but if someone wanted to point directly to a level URL with the anchor tag (to call attention to something via a callout at load time), it seemed like we should support it and not ignore it, so that it is a more officially supported part of the URL format.

> On Jun 9, 2016, at 2:29 PM, Brent Van Minnen notifications@github.com wrote:
> 
> In code-studio/src/js/callouts.js:
> 
> > ```
> >    $(selector).qtip(config).qtip('show');
> >  }
> > ```
> > 
> >    });
> > +
> > -  // Insert a hashchange handler to detect triggercallout= hashes and fire
> > -  // appropriate events to open the callout
> > -  function detectTriggerCalloutOnHash(event) {
> >   What is the scenario where the hash will be there on page load?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

Bjvanminnen replied:
It seems like one option would be to do something like https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/instructions/MarkdownInstructions.jsx#L59. In that case, after the component has mounted (i.e. there's a DOM representation) we look for any `details` elements and add some jquery handling to them.

We could also search for any descendants with an href containing "#triggercallout=" and for each instance we find, add a click event that triggers the appropriate name. The result would be that we don't have to listen to globally listen to hashchange events for this one feature, and we also dont need to worry about the current state of the hash (since the hash has just become a way to communicate to the instructions component what even it should trigger on click).

This would look something like:

```
$(ReactDOM.findDOMNode(this)).find("a[href*='#triggercallout=]").each((index, item) => {
   const href = item.attr('href');
   // extract trigger callout from href

   item.click(() => { $window.trigger(eventName, [eventType]); });
})
```

Also cc'ing @islemaster to see if he has any thoughts on the best way of accomplishing this, since he's good at coming up with clean ways of doing things :)

Regardless of which approach we take, this feature could also use some sort of tests. This feels like a code path that won't get a ton of coverage otherwise, and would be easy for us to regress and not realize.

cpirich replied:
I considered the "transform markdown at runtime" approach, but thought it would be nice to support this URL semantic directly (because jumping within the same document feels sort of like what hash URLs were intended for). Given that the scenario was "make it possible for HTML on the page to contain a link to jump to something within the same page", it felt like a good fit. I wouldn't want to use this as the normal method for communication on the page :) But if we ever did need to do something similar in the future, we could centralize one place to listen for the hashchange event, parse what's after the hash, and delegate as appropriate.

Regarding tests, are you thinking a UI test? that would cover most of the code and theoretically verify some of the toggling and scrolling behavior.

Bjvanminnen replied:
Yeah, I think UI test would probably be the easiest (only?) way to get coverage for this.

islemaster replied:
Ooh, this is an interesting feature, and I can see the reasoning behind both approaches.

In the current implementation we're treating the hash as an action, executed on page-load or on hash-change.  This is fairly consistent with the default behavior of hashes for jumping to anchors - it's a one time "scroll-to" action, and you're always free to scroll away after the action occurs.  In that sense, having a hash show a callout seems reasonable.

This is different from the way many webapps use hashes, as a sort of bookmarkable page-state: Imagine keeping an updated `#code` or `#design` in the URL to track which tab we're on in applab.  I don't recommend we ever take that approach, since we handle state differently, and that sort of thing tends to require more centralized routing than we are set up for.  I think react-router might be doing this under the hood though; maybe worth double-checking that we don't stomp on that somehow.

Brent's makes a good point about not depending on the `window.location`.  You could do postprocessing on instructions and even create a custom syntax for this sort of callout action that would read really nice in markdown:

```
You need to [draw sprites](callout:drawSpritesCallout) and
```

But there are downsides too:
- If you want to do a callout-showing link outside of instructions, you've got to do the postprocessing on whatever element contains the new link - and if you forget to do so, those links are broken!  With hashes the functionality is available to any hyperlink, anywhere.
- You lose the action-on-page-load feature.  You could always get it back by adding a queryparam for showing a callout on load, but then you've put more logic in another place.
- And of course, changing this means Chris has to tear out some stuff and write something else.  Not to suffer the sunk-cost fallacy, but that is a small factor.

Overall, I feel like sticking with the hash-action approach is fine for now, and could be revisited if we think it's conflicting with something.  I _would_ like to see the location and history manipulation extracted into some sort of `HashAction` module that callouts.js consumes, per SRP.

cpirich commented:
Adding example video
![callouts](https://cloud.githubusercontent.com/assets/5429146/15944497/bc65f03a-2e42-11e6-8e40-50c2419956d9.gif)

bcjordan commented:
That is awesome.

cpirich commented:
- Added support for `codeString` property in the callout definition. This will create a callout next to the line number where that piece of code appears using the `prepareforcallout` event. It works in both text mode (ace) and block mode (droplet) by first searching for the string (using ace's `find()` API). Then it takes advantage of a capability I put in earlier where the `selector` property can be modified in the callout's `codeStudio` object. It creates a new callout selector to attach to the appropriate line number (gutter) using a new pseudo-selector I added called `:textEquals` (required because the `nth-of-type` selector doesn't work once you scroll deep into a document - the editors only create a small number of gutter elements and they renumber them at runtime - so the selector needs to match on the line number text inside the element). And then the document is scrolled to that line number so the selector will match. These types of callouts now call `qtip('show')` asynchronously since DOM changes inside `prepareforcallout` don't occur immediately.

cpirich commented:
This video demonstrates this with a callout attached to the `codeString` of `// hello` - it finds that string and pops a callout in the appropriate place:
![callouts_codestring](https://cloud.githubusercontent.com/assets/5429146/15949185/52614bea-2e5a-11e6-9d3a-32f17820f27b.gif)

cpirich commented:
I'll add a UI test in a separate PR next week with additional changes to the `codeString` callouts to make them track and appear/disappear as code scrolls.
